### ISSUE TITLE
kernel-devel package have broken gcc plugin

### ISSUE THREAD

**Issue by marmarek** [QubesOS/qubes-issues#2844]
#### Qubes OS version (e.g., `R3.2`):

R3.2

#### Affected TemplateVMs (e.g., `fedora-23`, if applicable):

fedora-*

---

Since enabling gcc plugins, building kernel modules using kernel-devel package is broken:
```
make: Entering directory '/usr/src/kernels/4.9.29-17.pvops.qubes.x86_64'
  CC [M]  /var/tmp/u2mfn-3.2.4/u2mfn.o
cc1: error: cannot load plugin ./scripts/gcc-plugins/latent_entropy_plugin.so
   ./scripts/gcc-plugins/latent_entropy_plugin.so: undefined symbol: _Z16gsi_insert_afterP20gimple_stmt_iteratorP21gimple_statement_base19gsi_iterator_update
scripts/Makefile.build:299: recipe for target '/var/tmp/u2mfn-3.2.4/u2mfn.o' failed
make[1]: *** [/var/tmp/u2mfn-3.2.4/u2mfn.o] Error 1
Makefile:1490: recipe for target '_module_/var/tmp/u2mfn-3.2.4' failed
make: *** [_module_/var/tmp/u2mfn-3.2.4] Error 2
make: Leaving directory '/usr/src/kernels/4.9.29-17.pvops.qubes.x86_64'
```

This happen even if the kernel-devel package is built in exactly the same distribution, the same gcc version etc.

Kernel options responsible:
```
CONFIG_GCC_PLUGINS=y
# CONFIG_GCC_PLUGIN_CYC_COMPLEXITY is not set
CONFIG_GCC_PLUGIN_LATENT_ENTROPY=y
```

Those options were enabled in >=4.9.28-14, package now in current-testing:
https://github.com/QubesOS/updates-status/issues/55

/cc @rtiangha

**Comment by rtiangha** [comment]
This is ringing a bell, but I can't remember. It might be a missing dependency. When does this get triggered? Compiling u2mfn somewhere? I think I had to install elfutils-libelf-devel to fix it, and if that was the case, it'd need to be a Buildrequires for that kernel option but I might have forgotten to add it to kernel.spec. Can you try installing that package and try again?

**Comment by rtiangha** [comment]
And I think on Debian, I think it's called something different. libelf-dev, I believe.

**Comment by marmarek** [comment]
Steps to reproduce:
1. Get Fedora-based VM
2. Install qubes-vm-kernel-support
3. Download and install https://yum.qubes-os.org/r3.2/current-testing/dom0/fc23/rpm/kernel-devel-4.9.29-17.pvops.qubes.x86_64.rpm
4. Copy `/usr/src/u2mfn-3.2.4` somewhere (or switch to root)
5. Build that module with `make -C /usr/src/kernels/4.9.29-17.pvops.qubes.x86_64 M=$PWD module`

I've tried elfutils-libelf-devel but it doesn't help, neither gcc-plugins-devel.

**Comment by rtiangha** [comment]
This is on my dev machine. Is this the expected output?

    [user@KernelBuilder-fc23 tmp]$ cp -ar /usr/src/u2mfn-3.2.4 /tmp/
    [user@KernelBuilder-fc23 tmp]$ make -C /usr/src/kernels/4.9.29-17.pvops.qubes.x86_64 M=$PWD u2mfn-3.2.4/
    make: Entering directory '/usr/src/kernels/4.9.29-17.pvops.qubes.x86_64'
      LD      /tmp/u2mfn-3.2.4/built-in.o
      CC [M]  /tmp/u2mfn-3.2.4/u2mfn.o
    make: Leaving directory '/usr/src/kernels/4.9.29-17.pvops.qubes.x86_64'
    [user@KernelBuilder-fc23 tmp]$

If so, it probably is a dependency, but I'm not sure which one yet. I'll start a fresh template and try to find out.

My bash history cuts off so I'm not sure if I've caught everything I've installed in this template, but in the meantime, could it be one of these?

    coreutils module-init-tools sparse qubes-kernel-vm-support dracut busybox bc openssl-devel cmake
    debian-keyring debootstrap dpkg dpkg-dev dpkg-perl keyrings-filesystem perl-TimeDate
    ubu-keyring xz-lzma-compat yaml python-devel systemd-devel qubes-libvchan-xen-devel
    flex bison pycairo dracut rpm-sign sparse fedora-packager fedora-review

**Comment by marmarek** [comment]
I've re-checked template version and it looks like this is the problem: gcc plugin built on fc24 do work on fc24 and fc25 VM. But the one built on fc23 do not work on fc24+.
So, with gcc-plugin we no longer have universal kernel-devel package. Does it even worth pursuing that? There is always alternative of using pvgrub2 and installing kernel inside the (template)VM: https://www.qubes-os.org/doc/managing-vm-kernel/#using-kernel-installed-in-the-vm

**Comment by rtiangha** [comment]
Thanks; I just finished spinning up an FC25 template and was able to replicate your results.

Just throwing this out there: Would it work the opposite way? As in, compile all kernel related stuff in FC25 and then install that into FC23? As long as everything related to the kernel is compiled in FC25 and nothing gets recompiled in FC23, that *should* work as the kernel stuff should be able to work independently from the rest of the system. The only major issues would be if anyone had a need to compile their own kernel module (ex. custom driver) after the fact as there might be some issues with mismatched symbols due to the different versions of the compilers. But if no one needed to do that, maybe that could be an option?

Other options (maybe undesirable) are to update the FC23 gcc to match the one being used in FC25, although that may require other things to be recompiled too. Or another one could be to create a cross-compile toolchain in a Docker image that does only the kernel stuff and its output is what gets packaged as kernel updates for all the templates.

**Comment by rtiangha** [comment]
And just thinking out loud some more, if the issue is u2mfn getting recompiled at install time with a different compiler than the rest of the kernel was compiled through dracut or whatnot, I think you might be able to get around that by ensuring that a u2mfn module that was compiled with the same compiler the main kernel was compiled with is included in the package that gets installed (so, maybe create a kernel-qubes-modules package that has u2mfn and any other custom modules and ensure that gets installed alongside a corresponding kernel update and make sure those two get updated in tandem whenever there's a kernel update). That might be another option.

**Comment by marmarek** [comment]
u2mfn was just a convenient example - this is about other modules - for example if someone need out-of-tree wifi driver.

Doing `make gcc-plugins` in kernel headers dir (`/usr/src/kernels/.../`) fixes the issues. Needs `gcc-c++` and `gcc-plugin-devel` packages. But unfortunately it is incompatible with putting it to modules.img (https://github.com/rtiangha/qubes-linux-kernel/pull/2/commits/77f8e8ef6d6b1a804192906ba64c93f9c1ea899c), as it is read-only in VM.

**Comment by marmarek** [comment]
As for compiling the dom0 kernel under fc25 - I don't want to mix distributions/build environments that much. If anything, build it _again_ just for kernel-devel package - maybe we could skip actual build in that case.

**Comment by marmarek** [comment]
For now, the workflow for building kernel modules, with kernel headers included in modules.img is:
1. Copy `/lib/modules/$(uname -r)/build` somewhere:

       cp -r /lib/modules/$(uname -r)/build /tmp/build

2. Install required build dependencies: gcc-plugin-devel, gcc-c++ (and similar for Debian)
3. Re-build gcc plugin: 

       make -C /tmp/build gcc-plugins

4. Use `/tmp/build` for actual modules build: 

       make -C /tmp/build M=$PWD modules


This isn't exactly what I'd expect from provided kernel headers...

**Comment by Rudd-O** [comment]
This bug is literally preventing me from doing work at my workplace — and it's remained like this for what, three months now?  I'm trying to get a template's VMs to boot with an in-template kernel, and of course this doesn't work without the damn `u2mfn` module, which won't compile because of this exact bug.

Sigh.

**Comment by marmarek** [comment]
Be sure to compile it for that you compile u2mfn for that in-template kernel version - not for the one provided by dom0 - which fails because of this bug.

**Comment by Rudd-O** [comment]
I gave up trying to compile the u2mfn module, because the module automagically appears in /lib/modules/$kernel of appvm based on template (after setting the appvm to prgrub2).

Now I have a different problem, when I try to start any GUI app in the booted appvm, the GUI app simply does not start.

Quite disappointed that these simple things do not work at all.

**Comment by Rudd-O** [comment]
qubes-gui-agent.service dies with journal error Error while vchan send hdr, terminating.

I am stuck.

**Comment by Rudd-O** [comment]
And now (after a VM reboot) it doesn't hang anymore, but no xterm window appears when I qvm-run the xterm on the app VM.

**Comment by marmarek** [comment]
Those are exact symptoms of missing u2mfn module.

**Comment by Rudd-O** [comment]
But module is loaded.  It loads right after the login prompt in xl console of the VM, and lsmod confirms it is loaded.

**Comment by Rudd-O** [comment]
When are we abandoning the kernels in dom0 model in favor of a supported and well tested in-VM kernel support?

On August 28, 2017 4:31:51 PM GMT+02:00, Rudd-O <notifications@github.com> wrote:
>But module is loaded.  It loads right after the login prompt in xl
>console of the VM, and lsmod confirms it is loaded.
>
>-- 
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/QubesOS/qubes-issues/issues/2844#issuecomment-325370228

--

Rudd-O

http://rudd-o.com/

**Comment by flaviosanchez** [comment]
This issue should be changed from minor to major. You cannot compile Drivers, Crypto-Currency wallets and many other Applications! This is a shame because most people should assume that *INX users are the average user and will most likely be compiling a program or two at some point. I'll give two examples; I was able to compile PinkCoin wallet after following the instructions given here, but when I tried to run the application it failed. I had to get a developer from PinkCoin to send me their Ubuntu compiled file to run on Debian (plus I had to install other software like bootstrap). Then i also tried to compile lwfinger driver (https://github.com/lwfinger/rtl8188eu) without success. I also noticed you cannot make changes to this path "/usr/lib/modules/(uname -r)/kernel/drivers/net/wireless"

**Comment by flaviosanchez** [comment]
This issue should be changed from minor to major. This is a shame because most people should assume that *INX users are the average user and will most likely be compiling a program or two at some point. I'll give two examples; I was able to compile PinkCoin wallet after following the instructions given here, but when I tried to run the application it failed. I had to get a developer from PinkCoin to send me their Ubuntu compiled file to run in Debian (plus I had to install other software like bootstrap). Then i also tried to compile lwfinger driver to fix my WiFi issues (https://github.com/lwfinger/rtl8188eu). It failed to compile. I also noticed you cannot make changes to this path "/usr/lib/modules/(uname -r)/kernel/drivers/net/wireless"

**Comment by Skyr** [comment]
If it's of any use to someone: I wanted to install the wireguard module via dkms (thus the naming). The following script creates an rw overlay and is called via my /rw/config/rc.local:
```
#!/bin/bash
LOWER=/usr/lib/modules/$(uname -r)
UPPER=/rw/dkms-overlay/$(uname -r)
WORK=${UPPER}-work
rm -rf $WORK
mkdir -p $UPPER $WORK
mount -t overlay overlay -o lowerdir=${LOWER},upperdir=${UPPER},workdir=${WORK} ${LOWER}
```
For each new kernel, I do
`make -C /usr/lib/modules/$(uname -r)/build gcc-plugins` (fix gcc plugin)
`dkms install wireguard/0.0.20171221` (compile module)
Let's see if this works out… ;-)

**Comment by andrewdavidwong** [comment]
Reopening due to the regression reported in #5456.

**Comment by marmarek** [comment]
The issue was the fix was only in a master branch (kernel-latest-* packages), but not relevant stable branches. Fixed (4.19.82, 4.14.152).

**Comment by adrelanos** [comment]
I am still having this issue https://github.com/QubesOS/qubes-issues/issues/5456 despite using the kernel version which should have the fix for this.

dom0:

    dnf list | grep kernel

```
kernel.x86_64                          1000:4.19.43-1.pvops.qubes
kernel.x86_64                          1000:4.19.79-1.pvops.qubes
kernel.x86_64                          1000:4.19.94-1.pvops.qubes
kernel-qubes-vm.x86_64                 1000:4.19.43-1.pvops.qubes
kernel-qubes-vm.x86_64                 1000:4.19.79-1.pvops.qubes
kernel-qubes-vm.x86_64                 1000:4.19.94-1.pvops.qubes
qubes-core-dom0-linux-kernel-install.x86_64
```

```
DKMS make.log for lkrg-0.7 for kernel 4.19.0-5-amd64 (x86_64)
Sat 01 Feb 2020 11:07:41 AM EST
make -C /lib/modules/4.19.94-1.pvops.qubes.x86_64/build M=/var/lib/dkms/lkrg/0.7/build modules
make[1]: Entering directory '/lib/modules/4.19.94-1.pvops.qubes.x86_64/build'
  CC [M]  /var/lib/dkms/lkrg/0.7/build/src/modules/ksyms/p_resolve_ksym.o
  CC [M]  /var/lib/dkms/lkrg/0.7/build/src/modules/hashing/p_lkrg_fast_hash.o
  CC [M]  /var/lib/dkms/lkrg/0.7/build/src/modules/comm_channel/p_comm_channel.o
  CC [M]  /var/lib/dkms/lkrg/0.7/build/src/modules/integrity_timer/p_integrity_timer.o
  CC [M]  /var/lib/dkms/lkrg/0.7/build/src/modules/kmod/p_kmod.o
  CC [M]  /var/lib/dkms/lkrg/0.7/build/src/modules/database/CPU.o
  CC [M]  /var/lib/dkms/lkrg/0.7/build/src/modules/database/arch/x86/p_x86_metadata.o
  CC [M]  /var/lib/dkms/lkrg/0.7/build/src/modules/database/arch/x86/p_switch_idt/p_switch_idt.o
In file included from /var/lib/dkms/lkrg/0.7/build/src/modules/database/arch/x86/../../../../p_lkrg_main.h:30,
                 from /var/lib/dkms/lkrg/0.7/build/src/modules/database/arch/x86/p_x86_metadata.c:29:
./include/linux/random.h: In function ‘add_latent_entropy’:
./include/linux/random.h:26:39: error: ‘latent_entropy’ undeclared (first use in this function); did you mean ‘add_latent_entropy’?
  add_device_randomness((const void *)&latent_entropy,
                                       ^~~~~~~~~~~~~~
                                       add_latent_entropy
./include/linux/random.h:26:39: note: each undeclared identifier is reported only once for each function it appears in
In file included from /var/lib/dkms/lkrg/0.7/build/src/modules/database/arch/x86/p_switch_idt/../../../../../p_lkrg_main.h:30,
                 from /var/lib/dkms/lkrg/0.7/build/src/modules/database/arch/x86/p_switch_idt/p_switch_idt.c:24:
./include/linux/random.h: In function ‘add_latent_entropy’:
./include/linux/random.h:26:39: error: ‘latent_entropy’ undeclared (first use in this function); did you mean ‘add_latent_entropy’?
  add_device_randomness((const void *)&latent_entropy,
                                       ^~~~~~~~~~~~~~
                                       add_latent_entropy
./include/linux/random.h:26:39: note: each undeclared identifier is reported only once for each function it appears in
In file included from /var/lib/dkms/lkrg/0.7/build/src/modules/hashing/../../p_lkrg_main.h:30,
                 from /var/lib/dkms/lkrg/0.7/build/src/modules/hashing/p_lkrg_fast_hash.c:22:
./include/linux/random.h: In function ‘add_latent_entropy’:
./include/linux/random.h:26:39: error: ‘latent_entropy’ undeclared (first use in this function); did you mean ‘add_latent_entropy’?
  add_device_randomness((const void *)&latent_entropy,
                                       ^~~~~~~~~~~~~~
                                       add_latent_entropy
./include/linux/random.h:26:39: note: each undeclared identifier is reported only once for each function it appears in
In file included from /var/lib/dkms/lkrg/0.7/build/src/modules/ksyms/../../p_lkrg_main.h:30,
                 from /var/lib/dkms/lkrg/0.7/build/src/modules/ksyms/p_resolve_ksym.c:19:
./include/linux/random.h: In function ‘add_latent_entropy’:
./include/linux/random.h:26:39: error: ‘latent_entropy’ undeclared (first use in this function); did you mean ‘add_latent_entropy’?
  add_device_randomness((const void *)&latent_entropy,
                                       ^~~~~~~~~~~~~~
                                       add_latent_entropy
./include/linux/random.h:26:39: note: each undeclared identifier is reported only once for each function it appears in
In file included from /var/lib/dkms/lkrg/0.7/build/src/modules/kmod/../../p_lkrg_main.h:30,
                 from /var/lib/dkms/lkrg/0.7/build/src/modules/kmod/p_kmod.c:22:
./include/linux/random.h: In function ‘add_latent_entropy’:
./include/linux/random.h:26:39: error: ‘latent_entropy’ undeclared (first use in this function); did you mean ‘add_latent_entropy’?
  add_device_randomness((const void *)&latent_entropy,
                                       ^~~~~~~~~~~~~~
                                       add_latent_entropy
./include/linux/random.h:26:39: note: each undeclared identifier is reported only once for each function it appears in
In file included from /var/lib/dkms/lkrg/0.7/build/src/modules/integrity_timer/../../p_lkrg_main.h:30,
                 from /var/lib/dkms/lkrg/0.7/build/src/modules/integrity_timer/p_integrity_timer.c:18:
./include/linux/random.h: In function ‘add_latent_entropy’:
./include/linux/random.h:26:39: error: ‘latent_entropy’ undeclared (first use in this function); did you mean ‘add_latent_entropy’?
  add_device_randomness((const void *)&latent_entropy,
                                       ^~~~~~~~~~~~~~
                                       add_latent_entropy
./include/linux/random.h:26:39: note: each undeclared identifier is reported only once for each function it appears in
In file included from /var/lib/dkms/lkrg/0.7/build/src/modules/comm_channel/../../p_lkrg_main.h:30,
                 from /var/lib/dkms/lkrg/0.7/build/src/modules/comm_channel/p_comm_channel.c:18:
./include/linux/random.h: In function ‘add_latent_entropy’:
./include/linux/random.h:26:39: error: ‘latent_entropy’ undeclared (first use in this function); did you mean ‘add_latent_entropy’?
  add_device_randomness((const void *)&latent_entropy,
                                       ^~~~~~~~~~~~~~
                                       add_latent_entropy
./include/linux/random.h:26:39: note: each undeclared identifier is reported only once for each function it appears in
In file included from /var/lib/dkms/lkrg/0.7/build/src/modules/database/../../p_lkrg_main.h:30,
                 from /var/lib/dkms/lkrg/0.7/build/src/modules/database/CPU.c:44:
./include/linux/random.h: In function ‘add_latent_entropy’:
./include/linux/random.h:26:39: error: ‘latent_entropy’ undeclared (first use in this function); did you mean ‘add_latent_entropy’?
  add_device_randomness((const void *)&latent_entropy,
                                       ^~~~~~~~~~~~~~
                                       add_latent_entropy
./include/linux/random.h:26:39: note: each undeclared identifier is reported only once for each function it appears in
make[2]: *** [scripts/Makefile.build:303: /var/lib/dkms/lkrg/0.7/build/src/modules/database/arch/x86/p_switch_idt/p_switch_idt.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [scripts/Makefile.build:303: /var/lib/dkms/lkrg/0.7/build/src/modules/database/CPU.o] Error 1
make[2]: *** [scripts/Makefile.build:303: /var/lib/dkms/lkrg/0.7/build/src/modules/database/arch/x86/p_x86_metadata.o] Error 1
make[2]: *** [scripts/Makefile.build:303: /var/lib/dkms/lkrg/0.7/build/src/modules/kmod/p_kmod.o] Error 1
make[2]: *** [scripts/Makefile.build:303: /var/lib/dkms/lkrg/0.7/build/src/modules/ksyms/p_resolve_ksym.o] Error 1
make[2]: *** [scripts/Makefile.build:303: /var/lib/dkms/lkrg/0.7/build/src/modules/hashing/p_lkrg_fast_hash.o] Error 1
make[2]: *** [scripts/Makefile.build:303: /var/lib/dkms/lkrg/0.7/build/src/modules/comm_channel/p_comm_channel.o] Error 1
make[2]: *** [scripts/Makefile.build:303: /var/lib/dkms/lkrg/0.7/build/src/modules/integrity_timer/p_integrity_timer.o] Error 1
make[1]: *** [Makefile:1522: _module_/var/lib/dkms/lkrg/0.7/build] Error 2
make[1]: Leaving directory '/lib/modules/4.19.94-1.pvops.qubes.x86_64/build'
make: *** [Makefile:93: all] Error 2
```
